cooper/formulations/utils.py

```diff
@@ -85,61 +75,62 @@ def compute_dual_weighted_violation(
+    multiplier by :math:`\lambda` and the penalty coefficient by :math:`\rho`, we obtain
+    the contribution of an inequality constraint to the augmented Lagrangian:
 
     .. math::
-      \frac{1}{2 \rho} ( || max(0, \lambda + violation * rho) ||_2^2 - || \lambda ||_2^2)
+      \lambda_{*}^{\top} \text{violation} - \frac{1}{2 \rho} ||\lambda_{*} - \lambda||_2^2,
 
-    Note that when the multiplier is zero, the formula simplifies to the standard
-    quadratic penalty for inequality constraints.
-    .. math::
-        \frac{\rho}{2} || max(0, violation) ||_2^2
+    where :math:`\lambda_{*}= \texttt{relu}(\lambda + \rho \text{violation})`. Note that
+    this corresponds to the multiplier update after a step of projected gradient ascent.
 
-    When the constraint is an equality constraint, the quadratic penalty is computed
-    following Eq 17.36 in Numerical Optimization by Nocedal and Wright (2006). Note
-    that, unlike inequality constraints, there is no thresholding at zero for equality
-    constraints.
+    In the case of equality constraints, the quadratic-augmented contribution is computed
+    following Eq 17.36 in Numerical Optimization by Nocedal and Wright (2006):
 
     .. math::
-        \frac{rho}{2} ||violation||_2^2
+        \lambda^{\top} \text{violation}+ \frac{rho}{2} ||violation||_2^2
 
     """
 
-    if penalty_coefficient_value is None:
-        raise ValueError("The penalty coefficient tensor must be provided if the primal contribution is not skipped.")
-    if violation is None:
-        raise ValueError("The violation tensor must be provided if the primal contribution is not skipped.")
-
     if constraint_type == ConstraintType.INEQUALITY:
-
         aux1 = torch.einsum("i...,i...->i...", penalty_coefficient_value, violation)
-
-        if multiplier_value is None:
-            return 0.5 * torch.einsum("i...,i...->", 1 / penalty_coefficient_value, torch.relu(aux1) ** 2)
-        else:
-            aux2 = torch.relu(multiplier_value + aux1) ** 2 - multiplier_value**2
-            return 0.5 * torch.einsum("i...,i...->", 1 / penalty_coefficient_value, aux2)
-
+        aux2 = torch.relu(multiplier_value + aux1) ** 2 - multiplier_value**2
+        return 0.5 * torch.einsum("i...,i...->", 1 / penalty_coefficient_value, aux2)
     elif constraint_type == ConstraintType.EQUALITY:
-        if multiplier_value is None:
-            linear_contribution = 0.0
-        else:
-            linear_contribution = compute_primal_weighted_violation(
-                constraint_factor_value=multiplier_value, violation=violation
-            )
-        quadratic_penalty = 0.5 * torch.einsum("i...,i...->", penalty_coefficient_value, violation**2)
-        return linear_contribution + quadratic_penalty
+        # TODO(gallego-posada): Why is the linear term being computed with the "primal" weighted violation?
+        # Should we also have a "dual_weighted_violation"?
```

Comment:
Here we are computing the primal component of the augmented Lagrangian. It makes sense to use the primal weighted violation computation to (1) preserve gradients wrt the violations and (2) block gradients wrt the multipliers

Reply:
Consider renaming this method to indicate that it only deals with _primal_ contributions.
This would address the TODO